freqtrade/worker.py

```diff
@@ -0,0 +1,196 @@
+"""
+Main Freqtrade worker class.
+"""
+import logging
+import time
+import traceback
+from argparse import Namespace
+from typing import Any, Callable, Optional
+import sdnotify
+
+from freqtrade import (constants, OperationalException, TemporaryError,
+                       __version__)
+from freqtrade.configuration import Configuration
+from freqtrade.state import State
+from freqtrade.rpc import RPCMessageType
+
+
+logger = logging.getLogger(__name__)
+
+
+class Worker(object):
+    """
+    Freqtradebot worker class
+    """
+
+    def __init__(self, args: Namespace, config=None) -> None:
+        """
+        Init all variables and objects the bot needs to work
+        """
+        logger.info('Starting worker %s', __version__)
+
+        self._args = args
+        self._config = config
+        self._init(False)
+
+        # Tell systemd that we completed initialization phase
+        if self._sd_notify:
+            logger.debug("sd_notify: READY=1")
+            self._sd_notify.notify("READY=1")
+
+    def _init(self, reconfig: bool):
+        """
+        Also called from the _reconfigure() method (with reconfig=True).
+        """
+        if reconfig or self._config is None:
```

Comment:
I'm struggling with the `reconfig` parameter, since i'm not sure it's needed.

Apart from tests, this is only called twice, once from `__init__` - where it won't have a config yet - and from `_reconfigure()`, where you force loading the config.

I think in combination with moving state to freqtradebot - we use `get_patched_freqtrade` in `test_rpc_telegram.py` again, so the config-parameter in the constructor becomes irrelevant, since having config or not does not matter for the `worker` tests in `test_freqtradebot.py`.

Reply:
It's needed since the prepared config can be passed to the __init_() in the tests and therefore should not be created using args in the _init() member function.